src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs

```diff
@@ -959,6 +959,24 @@ private void ImportLdToken(int token)
                     }
                 }
 
+                // We might also be able to optimize this if this is something like typeof(Foo).IsValueType
+                // that doesn't need reflection metadata for the answer.
+                if (helperId != ReadyToRunHelperId.NecessaryTypeHandle)
+                {
+                    reader = new ILReader(_ilBytes, _currentOffset);
+                    if (reader.HasNext
+                        && reader.ReadILOpcode() == ILOpcode.call
+                        && IsTypeGetTypeFromHandle((MethodDesc)_methodIL.GetObject(reader.ReadILToken())))
+                    {
+                        if (reader.HasNext
+                            && reader.ReadILOpcode() is ILOpcode.callvirt or ILOpcode.call
+                            && _methodIL.GetObject(reader.ReadILToken()) is MethodDesc { Name: "get_IsValueType" or "get_IsEnum"})
```

Comment:
Add IsPrimitive too? It is used in some places in BCL.

Reply:
If we wanted to avoid coupling between the compiler and the BCL, we would make this pattern must-expand in the JIT. It would reduce the coupling to JIT/scanner that exists in many places.

Reply:
Do we have a test coverage for the (rare) situation when the JIT does not expand this pattern as an intrinsic?

Reply:
> If we wanted to avoid coupling between the compiler and the BCL, we would make this pattern must-expand in the JIT. It would reduce the coupling to JIT/scanner that exists in many places.

Hmm, I haven't even though about it in the context of the JIT optimization.

My original reason for why we don't need the metadata is because the implementation in `RuntimeType` will just go straight to the `MethodTable` if there is one and ask there. But yeah, RyuJIT will expand it before that, so we won't even get there.

I guess the only way to ensure this codepath really works would be to run unoptimized mode with the scanner enabled. I don't really know if I want to add a test pass with that. Or make it MustExpand in RyuJIT but not sure if that would be okay (cc @EgorBo).

Reply:
Not sure I understand how e.g. IsPrimitive can be a must-expand intrinsic -- ho do we expand it for `bool Foo(Type t) => t.IsPrimitive;` ?